src/Architectures.jl

```diff
@@ -25,23 +25,31 @@ variable `JULIA_NUM_THREADS` is set.
 struct CPU <: AbstractArchitecture end
 
 """
-    GPU <: AbstractArchitecture
+    GPU{D} <: AbstractArchitecture
 
-Run Oceananigans on a single NVIDIA CUDA GPU.
+Run Oceananigans on a single Nvidia CUDA or AMD ROCm GPU.
 """
-struct GPU <: AbstractArchitecture end
+struct GPU{D} <: AbstractArchitecture
+    device :: D
+end
+
+const CUDAGPU = GPU{<:CUDA.CUDABackend}
+
+GPU() = has_cuda() ? GPU(CUDA.CUDABackend()) : GPU(nothing)
```

Comment:
Let's add a docstring here. Also, I think we want to throw an error when no GPU is detected.

Reply:
@vchuravy can extensions overwrite methods? In other words, can `GPU()` throw an error if the user has not written `using CUDA`, but then automatically detect a CUDA GPU if they did write `using CUDA`?

We'd also possibly have to forbid writing both `using CUDA` and `using AMDGPU`. Or, have an extension that is only loaded when both of those packages are loaded? Is that possible?

Reply:
Maybe my dream of having autodetection upon calling `GPU()` is not possible...

Reply:
when we make both CUDA and AMDGPU extensions we could have e.g.  `__init()__` statements in either that would spit warning if they find that methods from the other extension are loaded